### ISSUE TITLE
Conditionals with a missing else-branch produce an unclear type error with let-expressions

### ISSUE THREAD

**Issue by dmbaturin** [ocaml/ocaml#10106]
If a conditional with a missing `else` branch contains only a simple boolean expression, then the compiler tells you what's wrong: missing `else` means implicit `unit`.

```ocaml
# if false then true ;;
Line 1, characters 14-18:
Error: This variant expression is expected to have type unit
         because it is in the result of a conditional with no else branch
       The constructor true does not belong to type unit

```
However, if you put a `let`-expression there, the compiler simply tells you that the expression is expected to have type `unit`, but no longer tells you the reason.

```ocaml
# 
if false then
  let () = print_endline "false is the new true" in
  true 
;;

Line 1, characters 64-68:
Error: This variant expression is expected to have type unit
       The constructor true does not belong to type unit

```

If would be nice if it could give a good error message in that situation too.

**Comment by gasche** [comment]
Thanks for the report! I proposed a fix in #10112.

**Comment by dmbaturin** [comment]
Thanks!
